crates/ruff_linter/src/rules/ruff/rules/suppression_comment_visitor.rs

```diff
@@ -0,0 +1,272 @@
+        self.parents.pop();
+
+        let node_end = node.end();
+
+        // Process all comments that start after the `preceding` node and end before this node's end.
+        while let Some(SuppressionComment { range, kind }) = self.comments.peek().copied() {
+            let line_position = CommentLinePosition::for_range(range, self.locator.contents());
+            if range.start() >= node_end {
+                if line_position.is_own_line() {
+                    if let Some(token) =
+                        SimpleTokenizer::starts_at(node_end, self.locator.contents())
+                            .skip_trivia()
+                            .next()
+                    {
+                        if token.end() <= range.end() {
+                            break;
+                        }
+                    }
+                    let comment_indent = indentation_at_offset(range.start(), self.locator)
+                        .unwrap_or_default()
+                        .len();
+                    let node_indent = indentation_at_offset(node.start(), self.locator)
+                        .unwrap_or_default()
+                        .len();
+                    if node_indent >= comment_indent {
+                        break;
+                    }
+                } else {
+                    if self.locator.line_start(range.start())
+                        != self.locator.line_start(node.start())
+                    {
+                        break;
+                    }
+                }
+            }
+
+            let data = SuppressionCommentData {
+                enclosing: Some(node),
+                preceding: self.preceding_node,
+                following: None,
+                line_position,
+                kind,
+                range,
+            };
+
+            self.builder.capture(data);
+            self.comments.next();
+        }
+
+        self.preceding_node = Some(node);
+    }
+    fn visit_body(&mut self, body: &'ast [ruff_python_ast::Stmt]) {
+        match body {
+            [] => {
+                // no-op
+            }
+            [only] => {
+                self.visit_stmt(only);
+            }
+            [first, .., last] => {
+                if self.can_skip(last.end()) {
+                    // Skip traversing the body when there's no comment between the first and last statement.
+                    // It is still necessary to visit the first statement to process all comments between
+                    // the previous node and the first statement.
+                    self.visit_stmt(first);
+                    self.preceding_node = Some(last.into());
+                } else {
+                    preorder::walk_body(self, body);
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(super) struct SuppressionCommentData<'src> {
+    /// If `enclosing` is `None`, this comment is top-level
+    pub(super) enclosing: Option<AnyNodeRef<'src>>,
```

Comment:
Nit: I think it would be helpful to explain the fields in more detail or refer to the formatter struct.